
![ImageName](https://cdn.kesci.com/upload/rudh9dbgjm.jpg)  

**作者：[lqy](https://www.heywhale.com/home/user/profile/5f27fd9633e1be002cc65a1d)** 华东师范大学气象学研究生、和鲸社区气象数据科学频道版主

🐋：本项目来自《[气象训练营⑦：WRF模式后处理](https://www.heywhale.com/home/competition/64478fec113e81a18dc70cd1)》活动，所有教案代码都可以一键跑通，你可以 fork 后在线在线运行、调试学习、完成作业练习。  

学习过程中如果你遇到任何问题，欢迎使用搜索引擎，或在 [讨论区](https://www.heywhale.com/home/competition/forumlist/64478fec113e81a18dc70cd1) 中发帖提出，我们很乐意为你提供帮助。

# 写在前面  

## 1. 活动背景    
    
+ WRF 模式是一个模拟天气过程的物理模式，作为一个 **免费开源**的气象模式，**高度灵活**，不仅在国内外气象业务和研究中被广泛使用，在地理、遥感、交通、环境、能源等领域也得到了广泛应用，能够与很多学科进行交叉研究。由于 WRF 模拟**数据格式特殊、维度高、体积大** 等，WRF 数据处理对于初学者有一定的难度。    
    
+ 因此，[本次训练营](https://www.heywhale.com/home/competition/64478fec113e81a18dc70cd1) 希望帮助大家 **更快速掌握 WRF 后处理技能，实现气象业务和科研的基本需求**，进一步挖掘气象数值模拟的价值。    
    
    
## 2. 学习内容    
    
+ WRF 后处理（Post-processing）是非常关键的一项技能，能够帮助我们更好地实现 **处理（Process）、可视化（Visualize）、比较（Compare）、分析（Analyze）、评估（Evaluate）** WRF 模拟结果。    
    
+ 本次训练营全部基于 Python 编程语言，利用 Python 工具 **（wrf-python、salem、xarray 和 cartopy 等）**实现 WRF 后处理，包含**数据读取、变量提取、物理量计算、时序分析、空间分布、时空演变等**，包括大量练习与闯关题，希望可以切实提高你的 WRF 后处理技术，更好地应对以后的科研与业务挑战。    
    
    
    
## 3. 基础预备        
    
    
+ 本次训练营提供了 **台风利奇马的 WRF 模拟结果（无需自己运行 WRF 模式）**。    
    
+ WRF 模式后处理需要一定的 **气象数据处理和可视化**的基本技能，建议提前熟悉**cartopy 和 xarray** 基础，非常推荐这两个训练营：    
    
	+ [超友好！气象Python训练营⑤xarray使用](https://www.heywhale.com/home/activity/detail/624408cb31153c001798cab2)    
    
	+ [超友好！气象Python训练营⑥绘图详解](https://www.heywhale.com/home/activity/detail/62c29fed0123ec8270e8a7b9/content/0)    



# 关卡一：解析 WRF 模式数据

作为第一关，我们需要**熟悉WRF模式数据的读取，并实现模拟变量的提取**，为实现WRF数据处理和可视化打下基础。

## 一、WRF 模式后处理（Post-processing）  

WRF模式后处理就是基于WRF模式的模拟结果进行处理，由于WRF模式是一个物理模型，我们需要先运行WRF模式，才能得到模拟结果，这也是为什么称之为**后处理**的原因。

WRF后处理与往常的数据处理流程是类似的，需要经过**读取数据→提取数据→处理数据→可视化数据**四个流程。但是，我们处理数据时更多是关注物理量的计算，可视化也需要在传统方式的基础上，实现高维空间数据可视化。

### （1）关于 WRF 模式

**WRF（Weather Research & Forecasting Model）**，是一个 **中尺度数值天气预报（Numeric Weather Prediction，NWP）** 模式，由美国环境预测中心（NCEP）和美国国家大气研究中心（NCAR）等科研机构中心联合研发，为大气研究和业务预报应用而设计。        
        
对于研究人员来说，WRF 可以根据实际的大气条件（即来自观测和分析）或理想化条件进行模拟，同时融合了很多专家学者在 **大气动力学、物理参数化方案和模式资料同化**方面的最新进展，从**中小尺度到全球尺度的数值预报和模拟** 都有广泛的应用 [^1]。.        
    
WRF 模式编译不仅需要一定的 Linux 基础，其运行也需要高性能的计算资源，造成了一定的使用门槛 [^2]。  

[^1]:如果想深入了解WRF模式，可以看下这三份材料：[WRF模式介绍](https://www.mmm.ucar.edu/models/wrf)、[WRF源码下载](https://github.com/wrf-model/WRF/releases)、[WRF使用手册](https://www2.mmm.ucar.edu/wrf/users/download/get_source.html)。  

[^2]:大家可以无痛点、不踩坑地在ModelWhale上进行WRF线上模拟。如果想尝试在线运行WRF模式，可以试试[Workshop|Python气象海洋数据分析II](https://www.heywhale.com/home/activity/detail/62de53410bbb187a397cbe76)第四期中[Hollis冬青老师](https://www.heywhale.com/home/user/profile/5f20f9b533e1be002cc5551f)发布的[中尺度天气预报模式WRF入门](https://www.heywhale.com/mw/project/628dc16d843e3c9fdc944a69)。 

### （2）WRF 数据预览工具

已有一些工具可以快速查看 WRF 数据变量，你可以在个人 PC 上安装后预览 WRF 模拟结果。这里推荐常用的3种工具：  
        
#### 1. Ncview （一般在支持图形显示的Linux系统上运行）  
> 使用手册：http://meteora.ucsd.edu/~pierce/ncview_home_page.html        
        
![ImageName](https://cdn.kesci.com/upload/image/rq9bamqua3.png)        
        
        
#### 2. Panoply （支持Linux、Windows和Mac系统，方便好用）  
> Panoply 官网：https://www.giss.nasa.gov/tools/panoply/        
        
![ImageName](https://cdn.kesci.com/upload/image/rq9awbiozb.png)        
      
#### 3. IDV  （支持Linux和Windows，有三维显示等高级功能）  
> IDV 官网：https://www.unidata.ucar.edu/software/idv/      
> 使用手册：https://docs.unidata.ucar.edu/idv/current/workshop/datadisplays/wrf/InteractiveWRFGridDisplays.html      
      
![ImageName](https://cdn.kesci.com/upload/image/rqxtl9bm24.png)      


### （3）WRF 数据处理工具

WRF 数据预览工具只能初步查看模拟结果，想要个性化地处理 WRF 数据，还是离不开自己动手编程。  
        
#### 1. NCL        
> 官网示例非常多，运行速度较快，绘图也很好看（但 NCL 已不再继续维护和开发）        
> 使用手册：https://www.ncl.ucar.edu/        
        
#### 2. wrf-python        
> 使用手册：https://wrf-python.readthedocs.io/en/latest/        
> 同样是 NCAR 官方开发，能够实现 NCL 的功能，对 Python 用户友好，兼容 xarray、cartopy 等模块        
        
#### 3. salem        
> 与 wrf-python 类似，有很多 WRF 数据处理功能，对 Python 用户友好，兼容 xarray、cartopy 等模块        
> 使用手册：https://salem.readthedocs.io/en/stable/index.html        
        
        
NCL 是一门相对古老的语言，使用起来较 Python 更为复杂。其实，熟练掌握 **wrf-python 和 salem** 能很大程度满足研究 / 业务需要。

### （4）WRF 模拟数据案例：模拟超强台风 “利奇马”

灾害性天气是大家非常关注的，台风的影响面积广且持续时间长，其风雨对我国沿海地区造成较大的威胁。因此，我们利用 WRF 4.2.2 版本模拟 **1909 号超强台风利奇马**，驱动场是 [NCEP0.25°FNL资料](https://rda.ucar.edu/datasets/ds083.2/)，积分开始时间是 2019 年 8 月 18 日 18 时（UTC）。数据集已挂载完毕，可以直接读取、使用。        
        
> [关于超强台风“利奇马”的一生](https://www.cma.gov.cn/2011xwzx/2011xqxxw/2011xzytq/201908/t20190814_533010.html)        
        
2019 年 8 月 10 日凌晨，“利奇马” 以超强台风姿态登陆浙江，中心附近最大风力 16 级（52 米 / 秒），一举成为 **1949 年以来登陆我国大陆地区第五强、登陆浙江第三强**。在浙江停滞 20 小时后，穿过江苏移入黄海，成为 **滞留浙江时间最长的超强台风**。        
        
超强台风 “利奇马”**两次登陆我国，两次入海**，最终在渤海缓缓结束一生。它 **登陆强度强、陆地滞留时间长、降雨强度大且极端性显著、大风影响范围广且持续时间长**，使华东及环渤海等地遭受严重风雨影响，浙江、安徽、江苏、山东等地均出现不同程度的城乡积涝、中小河流洪水、山洪和滑坡等灾害。        
        
        
![ImageName](https://cdn.kesci.com/upload/image/rq99nl2hna.png)        
        


## 二、熟悉 WRF 数据的文件名

In [1]:
!ls /home/mw/input/typhoon9537/  # 🐋：摁Ctrl/Shift+Enter即可运行本cell，不用谢

wrfout_d01_2019-08-08_18_00_00	wrfout_d01_2019-08-09_01_00_00
wrfout_d01_2019-08-08_19_00_00	wrfout_d01_2019-08-09_02_00_00
wrfout_d01_2019-08-08_20_00_00	wrfout_d01_2019-08-09_03_00_00
wrfout_d01_2019-08-08_21_00_00	wrfout_d01_2019-08-09_04_00_00
wrfout_d01_2019-08-08_22_00_00	wrfout_d01_2019-08-09_05_00_00
wrfout_d01_2019-08-08_23_00_00	wrfout_d01_2019-08-09_06_00_00
wrfout_d01_2019-08-09_00_00_00


WRF 输出数据的命名可以拆解为三个部分：        
        
1. WRF 模式输出数据都会以 `wrfout_` 开头        
        
2. `d01` 表示第一层网格（这里是单层模拟，只有 d01，如果是三层嵌套模拟，则有 d01、d02 和 d03）。        
        
3. 表示时间的字符串 `2019-08-08_18_00_00`，格式为 yyyy（年）\_mm（月）\_dd（日）\_hh（时）\_mm（分）\_ss（秒），注意这并不是北京时间，而是 **世界协调时（UTC）**。        
        
本次给出的 WRF 模拟结果是按逐小时输出的 [^1]。  

[^1]: 这个是在运行 WRF 时候自行设定的。

## 三、读取单个 WRF 模拟数据

虽然 wrfout 数据没有后缀格式，但 **本质上是 NETCDF 格式** 的数据。所以我们可以利用 netcdf4 或 xarray 模块读取 WRF 模拟数据、查看模拟设置。

### （1）基于 netCDF4 模块  

netcdf4-python 使用手册：https://unidata.github.io/netcdf4-python/

In [2]:
# 设置WRF数据目录
wrfout_path = '/home/mw/input/typhoon9537/'

In [3]:
from netCDF4 import Dataset

# 用netCDF4模块读取WRF模拟数据
ncfile = Dataset(wrfout_path + 'wrfout_d01_2019-08-08_18_00_00')
type(ncfile)

netCDF4._netCDF4.Dataset

In [4]:
# 查看基于netcdf4模块读取的变量个数
var_all = list(ncfile.variables.keys())
print(len(var_all))

166


##### ✍ 小练习：查看 WRF 模拟设置

In [5]:
# 查看ncfile的属性（由于返回结果较长，在此注释）
# 这里包含了WRF模拟设置（模拟范围、时间、分辨率、物理方案等等）
# print(ncfile)  # 🐋：可以摁 Ctrl+/ 取消注释

### （2）基于 xarray 模块

xarray 使用手册：https://docs.xarray.dev/en/stable/index.html  

xarray 不仅能读取netCDF4数据，还能实现多维数据的运算和可视化，是一个功能强大、应用广泛的气象数据处理模块。

In [6]:
import xarray as xr

# 用xarray包读取WRF模拟数据
ds = xr.open_dataset(wrfout_path + 'wrfout_d01_2019-08-08_18_00_00')
ds

<xarray.Dataset>
Dimensions:                (Time: 1, south_north: 437, west_east: 447, bottom_top: 34, bottom_top_stag: 35, soil_layers_stag: 4, west_east_stag: 448, south_north_stag: 438, seed_dim_stag: 2)
Coordinates:
    XLAT                   (Time, south_north, west_east) float32 ...
    XLONG                  (Time, south_north, west_east) float32 ...
    XTIME                  (Time) datetime64[ns] ...
    XLAT_U                 (Time, south_north, west_east_stag) float32 ...
    XLONG_U                (Time, south_north, west_east_stag) float32 ...
    XLAT_V                 (Time, south_north_stag, west_east) float32 ...
    XLONG_V                (Time, south_north_stag, west_east) float32 ...
Dimensions without coordinates: Time, south_north, west_east, bottom_top, bottom_top_stag, soil_layers_stag, west_east_stag, south_north_stag, seed_dim_stag
Data variables: (12/159)
    Times                  (Time) |S19 ...
    LU_INDEX               (Time, south_north, west_east) float32 ...
    ZNU                    (Time, bottom_top) float32 ...
    ZNW                    (Time, bottom_top_stag) float32 ...
    ZS                     (Time, soil_layers_stag) float32 ...
    DZS                    (Time, soil_layers_stag) float32 ...
    ...                     ...
    PCB                    (Time, south_north, west_east) float32 ...
    PC                     (Time, south_north, west_east) float32 ...
    LANDMASK               (Time, south_north, west_east) float32 ...
    LAKEMASK               (Time, south_north, west_east) float32 ...
    SST                    (Time, south_north, west_east) float32 ...
    SST_INPUT              (Time, south_north, west_east) float32 ...
Attributes: (12/134)
    TITLE:                            OUTPUT FROM WRF V4.4.2 MODEL
    START_DATE:                      2019-08-08_18:00:00
    SIMULATION_START_DATE:           2019-08-08_18:00:00
    WEST-EAST_GRID_DIMENSION:        448
    SOUTH-NORTH_GRID_DIMENSION:      438
    BOTTOM-TOP_GRID_DIMENSION:       35
    ...                              ...
    ISLAKE:                          21
    ISICE:                           15
    ISURBAN:                         13
    ISOILWATER:                      14
    HYBRID_OPT:                      2
    ETAC:                            0.2

In [7]:
# 查看基于xarray模块读取的变量个数
var_all = list(ds.keys())
print(len(var_all))

159


### （3）基于 salem 模块读取

> salem 使用手册：https://salem.readthedocs.io/en/stable/index.html        

第一次使用 `salem` 需要在 `home` 目录中创建 `.salem_cache` 文件夹，并下载 `salem-sample-data` 的数据包。为了更友好地直接导入 `salem` ，本项目已挂载社区的共享数据集[salem_cache文件集合](https://www.heywhale.com/mw/dataset/64716032b1251d62bfb38aa8/file)，再把该数据集的数据复制到 `/home/mw/.salem_cache/`  内，实现手动下载，避免卡住。

In [8]:
import os
import os.path

if(not os.path.exists('./salem_cache')):
    os.mkdir('./salem_cache') # 创建后它隐形的

! cp -r /home/mw/input/salem_cache05279422/salem_cache/ /home/mw/.salem_cache/ # 拷贝文件
! ls -ll /home/mw/.salem_cache/  # 检查有无成功

total 52928
drwxr-xr-x 3 mw users     4096 May 29 03:30 cache
drwxr-xr-x 2 mw users     4096 May 29 03:30 downloads
drwxr-xr-x 5 mw users     4096 May 29 03:31 salem_cache
drwxr-xr-x 6 mw users     4096 May 29 03:30 salem-sample-data-758f7ddd0fa6b5b1bd4c63b6dcfe8d5eec0f4c59
-rw-r--r-- 1 mw users 54177895 May 29 03:30 salem-sample-data-758f7ddd0fa6b5b1bd4c63b6dcfe8d5eec0f4c59.zip


In [9]:
import salem
# 用salem包读取数据
ds = salem.open_wrf_dataset(wrfout_path + 'wrfout_d01_2019-08-08_18_00_00')
ds

<xarray.Dataset>
Dimensions:                (south_north: 437, west_east: 447, time: 1, bottom_top: 34, soil_layers: 3, seed_dim: 1)
Coordinates:
    lat                    (south_north, west_east) float32 20.89 ... 32.78
    lon                    (south_north, west_east) float32 116.5 ... 130.3
    xtime                  (time) datetime64[ns] ...
  * time                   (time) datetime64[ns] 2019-08-08T18:00:00
  * west_east              (west_east) float64 -6.69e+05 -6.66e+05 ... 6.69e+05
  * south_north            (south_north) float64 -6.54e+05 ... 6.54e+05
Dimensions without coordinates: bottom_top, soil_layers, seed_dim
Data variables: (12/166)
    LU_INDEX               (time, south_north, west_east) float32 ...
    ZNU                    (time, bottom_top) float32 ...
    ZNW                    (time, bottom_top) float32 ...
    ZS                     (time, soil_layers) float32 ...
    DZS                    (time, soil_layers) float32 ...
    VAR_SSO                (time, south_north, west_east) float32 ...
    ...                     ...
    TK                     (time, bottom_top, south_north, west_east) float32 ...
    WS                     (time, bottom_top, south_north, west_east) float32 ...
    PRESSURE               (time, bottom_top, south_north, west_east) float32 ...
    GEOPOTENTIAL           (time, bottom_top, south_north, west_east) float32 ...
    Z                      (time, bottom_top, south_north, west_east) float32 ...
    SLP                    (time, south_north, west_east) float32 ...
Attributes: (12/135)
    TITLE:                            OUTPUT FROM WRF V4.4.2 MODEL
    START_DATE:                      2019-08-08_18:00:00
    SIMULATION_START_DATE:           2019-08-08_18:00:00
    WEST-EAST_GRID_DIMENSION:        448
    SOUTH-NORTH_GRID_DIMENSION:      438
    BOTTOM-TOP_GRID_DIMENSION:       35
    ...                              ...
    ISICE:                           15
    ISURBAN:                         13
    ISOILWATER:                      14
    HYBRID_OPT:                      2
    ETAC:                            0.2
    pyproj_srs:                      +proj=lcc +lat_0=27 +lon_0=123 +lat_1=20...

In [10]:
# 查看基于salem模块读取的变量个数
var_all = list(ds.keys())
print(len(var_all))

166


对比下 netcdf4 读取出来的变量结果，让我们做一个小练习。

#### ✍ 小练习：比较基于 xarray 和 salem 模块读取的 WRF 模拟数据变量差异

我们分别用 netcdf4、xarray 和 salem 模块读取的 WRF 模拟数据变量名（列表 `var_all`），可以看到 netcdf4 和 salem 模块读取的变量为 166 个，而 xarray 读取的变量为 159 个，你能找到它们之间的差异吗？  

提示：将变量名称添加到列表，并计算两个列表的差集

## 四、合并多个 WRF 模拟数据

通常一次 WRF 模拟会输出很多时次的数据，为了更好地分析时序变化，需要将所需模拟时次的 WRF 数据进行合并。

### 会失败的尝试 1：基于 xarray 模块读取

In [11]:
# 尝试1：基于xarray模块无法读取  
# 传统的nc数据是可以根据时间维度拼接的，但是对WRF模拟数据并不支持  
# ds = xr.open_mfdataset(wrfout_path + 'wrfout_d01_*')  

### 不推荐的尝试 2：使用循环读取内存溢出

虽然 xarray 的合并读取报错了，于是有同学会想到用自己构建一个列表，然后按照指定的维度（Time）进行合并。        


In [12]:
# from glob import glob        
# # 得到wrfout数据列表        
# wrf_list = sorted(glob(wrfout_path + 'wrfout_*'))        
        
# # 循环的方式将xarray读取结果添加到ds_list        
# ds_list = []        
# for i in wrf_list:        
#     ds = xr.open_dataset(i)        
#     ds_list.append(ds)        
        
# # 按时间维度合并数据        
# ds_all = xr.concat(ds_list, "Time")        
# ds_all   


        
        
**这个方法非常不推荐！！因为 WRF 数据比较大，非常容易超出内存限制！！**

### 正解：salem 模块一行搞定

最简单高效的方法：使用`salem.open_mf_wrf_dataset`函数

In [13]:
import salem
# 用salem模块进行合并读取
ds_all = salem.open_mf_wrf_dataset(wrfout_path + 'wrfout_d01_*')
ds_all

<xarray.Dataset>
Dimensions:                (south_north: 437, west_east: 447, time: 13, bottom_top: 34, soil_layers: 3, seed_dim: 1)
Coordinates:
    lat                    (south_north, west_east) float32 20.89 ... 32.78
    lon                    (south_north, west_east) float32 116.5 ... 130.3
    xtime                  (time) datetime64[ns] dask.array<chunksize=(1,), meta=np.ndarray>
  * time                   (time) datetime64[ns] 2019-08-08T18:00:00 ... 2019...
  * west_east              (west_east) float64 -6.69e+05 -6.66e+05 ... 6.69e+05
  * south_north            (south_north) float64 -6.54e+05 ... 6.54e+05
Dimensions without coordinates: bottom_top, soil_layers, seed_dim
Data variables: (12/166)
    LU_INDEX               (time, south_north, west_east) float32 dask.array<chunksize=(1, 437, 447), meta=np.ndarray>
    ZNU                    (time, bottom_top) float32 dask.array<chunksize=(1, 34), meta=np.ndarray>
    ZNW                    (time, bottom_top) float32 dask.array<chunksize=(1, 34), meta=np.ndarray>
    ZS                     (time, soil_layers) float32 dask.array<chunksize=(1, 3), meta=np.ndarray>
    DZS                    (time, soil_layers) float32 dask.array<chunksize=(1, 3), meta=np.ndarray>
    VAR_SSO                (time, south_north, west_east) float32 dask.array<chunksize=(1, 437, 447), meta=np.ndarray>
    ...                     ...
    TK                     (time, bottom_top, south_north, west_east) float32 dask.array<chunksize=(1, 34, 437, 447), meta=np.ndarray>
    WS                     (time, bottom_top, south_north, west_east) float32 dask.array<chunksize=(1, 34, 437, 447), meta=np.ndarray>
    PRESSURE               (time, bottom_top, south_north, west_east) float32 dask.array<chunksize=(1, 34, 437, 447), meta=np.ndarray>
    GEOPOTENTIAL           (time, bottom_top, south_north, west_east) float32 dask.array<chunksize=(1, 34, 437, 447), meta=np.ndarray>
    Z                      (time, bottom_top, south_north, west_east) float32 dask.array<chunksize=(1, 34, 437, 447), meta=np.ndarray>
    SLP                    (time, south_north, west_east) float32 dask.array<chunksize=(1, 437, 447), meta=np.ndarray>
Attributes: (12/135)
    TITLE:                            OUTPUT FROM WRF V4.4.2 MODEL
    START_DATE:                      2019-08-08_18:00:00
    SIMULATION_START_DATE:           2019-08-08_18:00:00
    WEST-EAST_GRID_DIMENSION:        448
    SOUTH-NORTH_GRID_DIMENSION:      438
    BOTTOM-TOP_GRID_DIMENSION:       35
    ...                              ...
    ISICE:                           15
    ISURBAN:                         13
    ISOILWATER:                      14
    HYBRID_OPT:                      2
    ETAC:                            0.2
    pyproj_srs:                      +proj=lcc +lat_0=27 +lon_0=123 +lat_1=20...

这个函数读取不仅代码简单，而且效率高，是因为其采用了并行读取的方式，并没有直接将全部数据加载进内存。感兴趣可以与上述循环读取方式进行比对。

## 五、WRF 模拟变量的提取（含合并提取）

由于使用了不同的读取方式，xarray 和 salem 模块读取的返回结果都是 xarray.Dataset 格式，而 netCDF4 读取的返回结果是 netCDF4.\_netCDF4.Dataset 格式。因此，变量提取方法上也会存在一些差异。

### 读取方法一：使用 xarray/salem  

友情提醒：对 FNL 数据结构不熟悉的同学，可以看看这篇《[用python轻松提取grib文件数据，只用3个步骤](https://www.heywhale.com/mw/project/62537e970b30350017514b5d)》。

In [14]:
# 提取特定变量，以海平面气压为例，找到变量名称即可访问数据
ds_all.SLP

<xarray.DataArray 'SLP' (time: 13, south_north: 437, west_east: 447)>
dask.array<concatenate, shape=(13, 437, 447), dtype=float32, chunksize=(1, 437, 447)>
Coordinates:
    lat          (south_north, west_east) float32 20.89 20.89 ... 32.78 32.78
    lon          (south_north, west_east) float32 116.5 116.6 ... 130.2 130.3
    xtime        (time) datetime64[ns] dask.array<chunksize=(1,), meta=np.ndarray>
  * time         (time) datetime64[ns] 2019-08-08T18:00:00 ... 2019-08-09T06:...
  * west_east    (west_east) float64 -6.69e+05 -6.66e+05 ... 6.66e+05 6.69e+05
  * south_north  (south_north) float64 -6.54e+05 -6.51e+05 ... 6.51e+05 6.54e+05
Attributes:
    FieldType:    104
    MemoryOrder:  XY 
    description:  Sea level pressure
    units:        hPa
    stagger:      
    pyproj_srs:   +proj=lcc +lat_0=27 +lon_0=123 +lat_1=20 +lat_2=40 +x_0=0 +...

In [15]:
# 筛选特定的时间范围
ds_all.SLP.sel(time=slice('2019-08-08 18:00:00', '2019-08-09 00:00:00'))

<xarray.DataArray 'SLP' (time: 7, south_north: 437, west_east: 447)>
dask.array<getitem, shape=(7, 437, 447), dtype=float32, chunksize=(1, 437, 447)>
Coordinates:
    lat          (south_north, west_east) float32 20.89 20.89 ... 32.78 32.78
    lon          (south_north, west_east) float32 116.5 116.6 ... 130.2 130.3
    xtime        (time) datetime64[ns] dask.array<chunksize=(1,), meta=np.ndarray>
  * time         (time) datetime64[ns] 2019-08-08T18:00:00 ... 2019-08-09
  * west_east    (west_east) float64 -6.69e+05 -6.66e+05 ... 6.66e+05 6.69e+05
  * south_north  (south_north) float64 -6.54e+05 -6.51e+05 ... 6.51e+05 6.54e+05
Attributes:
    FieldType:    104
    MemoryOrder:  XY 
    description:  Sea level pressure
    units:        hPa
    stagger:      
    pyproj_srs:   +proj=lcc +lat_0=27 +lon_0=123 +lat_1=20 +lat_2=40 +x_0=0 +...

### 读取方法二：使用 netCDF4

`wrf-python` 模块中的 `getvar` 函数可以提取 netCDF4 读取的结果 [^1]。此处getvar函数与之前xarray模块提取的结果不会存在数值上的差异。  
        
[^1]: `getvar` 可以提取的 WRF 变量：https://wrf-python.readthedocs.io/en/latest/user_api/generated/wrf.getvar.html

In [16]:
from wrf import getvar
# 用netCDF4包读取WRF模拟数据
wrf_file = Dataset(wrfout_path + 'wrfout_d01_2019-08-08_18_00_00')
# timeidx可以表示时间序列的顺序，可为正（顺序索引），可为负（倒序索引）
slp = getvar(wrf_file, 'slp', timeidx=-1)
slp

<xarray.DataArray 'slp' (south_north: 437, west_east: 447)>
array([[ 996.07623,  996.0524 ,  996.0279 , ..., 1000.29803, 1000.34375,
        1000.3886 ],
       [ 996.06226,  996.0388 ,  996.01324, ..., 1000.278  , 1000.32275,
        1000.3663 ],
       [ 996.04767,  996.02386,  995.9985 , ..., 1000.25854, 1000.3015 ,
        1000.34485],
       ...,
       [1003.345  , 1003.3418 , 1003.34296, ..., 1007.48126, 1007.6548 ,
        1007.89465],
       [1003.3717 , 1003.36755, 1003.3666 , ..., 1007.5617 , 1007.6869 ,
        1008.16376],
       [1003.398  , 1003.393  , 1003.3914 , ..., 1007.5654 , 1007.7246 ,
        1008.1277 ]], dtype=float32)
Coordinates:
    XLONG    (south_north, west_east) float32 116.5 116.6 116.6 ... 130.2 130.3
    XLAT     (south_north, west_east) float32 20.89 20.89 20.89 ... 32.78 32.78
    XTIME    float32 0.0
    Time     datetime64[ns] 2019-08-08T18:00:00
Dimensions without coordinates: south_north, west_east
Attributes:
    FieldType:    104
    MemoryOrder:  XY
    description:  sea level pressure
    units:        hPa
    stagger:      
    coordinates:  XLONG XLAT XTIME
    projection:   LambertConformal(stand_lon=123.0, moad_cen_lat=27.0, truela...

In [17]:
# 如果想将xarray.DataArray()换成更为熟悉的numpy.array()，可以设置禁用xarray
# 方法一：在getvar函数中添加meta参数，并设置为False
slp_np_1 = getvar(wrf_file, 'slp', timeidx=-1, meta=False)
# 方法二：使用to_np函数
from wrf import to_np
slp_np_2 = to_np(slp)

### 合并提取方法一：利用 cat 方法  

有时候，我们为了避免单个WRF模拟数据存储过大，会将其分为多个时次的数据输出。这样虽然减少了单个数据的存储大小，但是对时间连续分析造成了不便，因此需要有一个根据时间来合并提取的方法。  

下面这个案例展示了合并提取多个时次的海平面气压数据。

In [18]:
from wrf import getvar, ALL_TIMES
import os

# 构建一个WRF数据列表
'''
wrf_list = [Dataset(wrfout_path + "wrfout_d01_2019-08-08_18_00_00"),
            Dataset(wrfout_path + "wrfout_d01_2019-08-08_19_00_00"),
            Dataset(wrfout_path + "wrfout_d01_2019-08-08_20_00_00"),
            Dataset(wrfout_path + "wrfout_d01_2019-08-08_21_00_00"),
            Dataset(wrfout_path + "wrfout_d01_2019-08-08_22_00_00"),
            Dataset(wrfout_path + "wrfout_d01_2019-08-08_23_00_00"),
            Dataset(wrfout_path + "wrfout_d01_2019-08-09_00_00_00")]
'''

wrf_list = [Dataset(wrfout_path + f)
            for f in os.listdir(wrfout_path) if f.startswith("wrfout_d01_")]

# 利用cat方法，提取所有模拟时段的海平面气压（slp）数据
# cat方法会将序列中所有文件沿着 'Time' 维进行合并，时间维度将作为返回数组的最左侧维度。
# 为了在输出数组中包含所有文件中的所有时间，设置 timeidx 参数为 wrf.ALL_TIMES（或设置为 None）。
# 如果 timeidx 是单个值，那么将假设时间索引取自所有文件所有时间的连接。

slp = getvar(wrf_list, 'slp', timeidx=ALL_TIMES, method='cat')
slp

<xarray.DataArray 'slp' (Time: 13, south_north: 437, west_east: 447)>
array([[[ 996.7818 ,  996.75543,  996.72644, ..., 1001.1558 ,
         1001.21   , 1001.247  ],
        [ 996.773  ,  996.70154,  996.6699 , ..., 1001.1916 ,
         1001.19904, 1001.19855],
        [ 996.73035,  996.6652 ,  996.6341 , ..., 1001.18945,
         1001.2004 , 1001.2167 ],
        ...,
        [1003.20624, 1003.1967 , 1003.1872 , ..., 1007.455  ,
         1008.0026 , 1008.0779 ],
        [1003.2347 , 1003.22925, 1003.2153 , ..., 1007.42664,
         1007.8822 , 1008.3424 ],
        [1003.2472 , 1003.26154, 1003.2455 , ..., 1007.8368 ,
         1007.9789 , 1008.2763 ]],

       [[ 996.7313 ,  996.72107,  996.7123 , ..., 1000.9793 ,
         1001.01917, 1001.0587 ],
        [ 996.6876 ,  996.6812 ,  996.68774, ..., 1000.9455 ,
         1000.97217, 1001.0505 ],
        [ 996.6451 ,  996.66016,  996.6751 , ..., 1000.93854,
         1000.98   , 1001.04224],
...
        [1002.12573, 1002.0649 , 1002.0236 , ..., 1006.11475,
         1006.92664, 1007.1496 ],
        [1002.12994, 1002.1158 , 1002.0759 , ..., 1006.0648 ,
         1006.8177 , 1007.46484],
        [1002.1822 , 1002.1928 , 1002.13153, ..., 1006.59424,
         1006.9374 , 1007.43475]],

       [[ 996.26514,  996.2407 ,  996.1901 , ..., 1000.50555,
         1000.5537 , 1000.5964 ],
        [ 996.22766,  996.19403,  996.16754, ..., 1000.52747,
         1000.54846, 1000.59247],
        [ 996.2059 ,  996.17035,  996.1449 , ..., 1000.53094,
         1000.54095, 1000.53094],
        ...,
        [1003.3169 , 1003.2893 , 1003.2703 , ..., 1007.293  ,
         1007.8166 , 1007.92175],
        [1003.3373 , 1003.313  , 1003.3137 , ..., 1007.2611 ,
         1007.6963 , 1008.20807],
        [1003.3665 , 1003.3415 , 1003.3366 , ..., 1007.62726,
         1007.7819 , 1008.1645 ]]], dtype=float32)
Coordinates:
    XLONG     (south_north, west_east) float32 116.5 116.6 116.6 ... 130.2 130.3
    XLAT      (south_north, west_east) float32 20.89 20.89 20.89 ... 32.78 32.78
    XTIME     (Time) float64 240.0 720.0 300.0 480.0 ... 660.0 180.0 600.0 60.0
  * Time      (Time) datetime64[ns] 2019-08-08T22:00:00 ... 2019-08-08T19:00:00
    datetime  (Time) datetime64[ns] 2019-08-08T22:00:00 ... 2019-08-08T19:00:00
Dimensions without coordinates: south_north, west_east
Attributes:
    FieldType:    104
    MemoryOrder:  XY
    description:  sea level pressure
    units:        hPa
    stagger:      
    coordinates:  XLONG XLAT XTIME
    projection:   LambertConformal(stand_lon=123.0, moad_cen_lat=27.0, truela...

### 合并提取方法二：利用 join 方法  

+ 使用join方法合并一系列文件时，会将文件/序列索引作为新数组的最左侧维度。  
+ 当有多个文件并且每个文件具有多个时间时，如果最后一个文件的时间数少于之前文件的时间数，那么剩余的数组将用缺省值填充。

In [19]:
# 构建一个WRF数据列表
wrf_list = [Dataset(wrfout_path + f)
            for f in os.listdir(wrfout_path) if f.startswith("wrfout_d01_")]

# 利用join方法，提取所有模拟时段的海平面气压（slp）数据
# 虽然本次案例中join方法和cat方法效果一样，但大部分情况下，只需要使用 cat 方法即可。

slp = getvar(wrf_list, 'slp', timeidx=ALL_TIMES, method='join')
slp

<xarray.DataArray 'slp' (file: 13, south_north: 437, west_east: 447)>
array([[[ 996.7818 ,  996.75543,  996.72644, ..., 1001.1558 ,
         1001.21   , 1001.247  ],
        [ 996.773  ,  996.70154,  996.6699 , ..., 1001.1916 ,
         1001.19904, 1001.19855],
        [ 996.73035,  996.6652 ,  996.6341 , ..., 1001.18945,
         1001.2004 , 1001.2167 ],
        ...,
        [1003.20624, 1003.1967 , 1003.1872 , ..., 1007.455  ,
         1008.0026 , 1008.0779 ],
        [1003.2347 , 1003.22925, 1003.2153 , ..., 1007.42664,
         1007.8822 , 1008.3424 ],
        [1003.2472 , 1003.26154, 1003.2455 , ..., 1007.8368 ,
         1007.9789 , 1008.2763 ]],

       [[ 996.7313 ,  996.72107,  996.7123 , ..., 1000.9793 ,
         1001.01917, 1001.0587 ],
        [ 996.6876 ,  996.6812 ,  996.68774, ..., 1000.9455 ,
         1000.97217, 1001.0505 ],
        [ 996.6451 ,  996.66016,  996.6751 , ..., 1000.93854,
         1000.98   , 1001.04224],
...
        [1002.12573, 1002.0649 , 1002.0236 , ..., 1006.11475,
         1006.92664, 1007.1496 ],
        [1002.12994, 1002.1158 , 1002.0759 , ..., 1006.0648 ,
         1006.8177 , 1007.46484],
        [1002.1822 , 1002.1928 , 1002.13153, ..., 1006.59424,
         1006.9374 , 1007.43475]],

       [[ 996.26514,  996.2407 ,  996.1901 , ..., 1000.50555,
         1000.5537 , 1000.5964 ],
        [ 996.22766,  996.19403,  996.16754, ..., 1000.52747,
         1000.54846, 1000.59247],
        [ 996.2059 ,  996.17035,  996.1449 , ..., 1000.53094,
         1000.54095, 1000.53094],
        ...,
        [1003.3169 , 1003.2893 , 1003.2703 , ..., 1007.293  ,
         1007.8166 , 1007.92175],
        [1003.3373 , 1003.313  , 1003.3137 , ..., 1007.2611 ,
         1007.6963 , 1008.20807],
        [1003.3665 , 1003.3415 , 1003.3366 , ..., 1007.62726,
         1007.7819 , 1008.1645 ]]], dtype=float32)
Coordinates:
    XLONG     (south_north, west_east) float32 116.5 116.6 116.6 ... 130.2 130.3
    XLAT      (south_north, west_east) float32 20.89 20.89 20.89 ... 32.78 32.78
    XTIME     (file) float32 240.0 720.0 300.0 480.0 ... 660.0 180.0 600.0 60.0
  * file      (file) int64 0 1 2 3 4 5 6 7 8 9 10 11 12
    datetime  (file) datetime64[ns] 2019-08-08T22:00:00 ... 2019-08-08T19:00:00
Dimensions without coordinates: south_north, west_east
Attributes:
    FieldType:    104
    MemoryOrder:  XY
    description:  sea level pressure
    units:        hPa
    stagger:      
    coordinates:  XLONG XLAT XTIME
    projection:   LambertConformal(stand_lon=123.0, moad_cen_lat=27.0, truela...

### ✍ 小练习：从多时次数据中提取多个变量

我们学习了如何从多个 wrfout 数中提取单个变量，但有时候我们想同时提取多个变量。尝试提取海平面气压和 2m 温度这两个变量并写入到同一个文件。

In [20]:
# WRF数据列表
wrf_list = [Dataset(wrfout_path + f)
            for f in os.listdir(wrfout_path) if f.startswith("wrfout_d01_")]
# 提取海平面气压数据
slp = getvar(wrf_list, 'slp', timeidx=ALL_TIMES, method='cat')
# 提取2m温度数据
t2m = getvar(wrf_list, 'T2', timeidx=ALL_TIMES, method='cat')

# 合并海平面气压和2m温度数据（你的代码）



### ✍ 小练习：提取 WRF 数据的时间列表

通过 `extract_times` 函数，我们可以提取 WRF 数据对应的时间。

In [21]:
# from wrf import extract_times
# # 从WRF数据列表中提取时间列表
# times_all = extract_times(wrf_list, timeidx=ALL_TIMES)
# times_all

## 六、WRF 数据读取的高级功能

如果我们有多套方案的 WRF 模拟数据，通常会存放在不同的文件夹内。这时候可以采用如下方式，在合并 WRF 数据的时候，再增加一个维度，来表示我们设定的模拟方案。

In [22]:
# case1表示实验方案1，case2表示实验方案2，……
wrf_dict = {"case1": [Dataset(wrfout_path + "wrfout_d01_2019-08-08_18_00_00"),
                      Dataset(wrfout_path + "wrfout_d01_2019-08-08_19_00_00"),
                      Dataset(wrfout_path + "wrfout_d01_2019-08-08_20_00_00")],
            "case2": [Dataset(wrfout_path + "wrfout_d01_2019-08-08_18_00_00"),
                      Dataset(wrfout_path + "wrfout_d01_2019-08-08_19_00_00"),
                      Dataset(wrfout_path + "wrfout_d01_2019-08-08_20_00_00")]
            }

slp = getvar(wrf_dict, 'slp', timeidx=ALL_TIMES)
slp

<xarray.DataArray 'slp' (key_0: 2, Time: 3, south_north: 437, west_east: 447)>
array([[[[ 996.07623,  996.0524 ,  996.0279 , ..., 1000.29803,
          1000.34375, 1000.3886 ],
         [ 996.06226,  996.0388 ,  996.01324, ..., 1000.278  ,
          1000.32275, 1000.3663 ],
         [ 996.04767,  996.02386,  995.9985 , ..., 1000.25854,
          1000.3015 , 1000.34485],
         ...,
         [1003.345  , 1003.3418 , 1003.34296, ..., 1007.48126,
          1007.6548 , 1007.89465],
         [1003.3717 , 1003.36755, 1003.3666 , ..., 1007.5617 ,
          1007.6869 , 1008.16376],
         [1003.398  , 1003.393  , 1003.3914 , ..., 1007.5654 ,
          1007.7246 , 1008.1277 ]],

        [[ 996.26514,  996.2407 ,  996.1901 , ..., 1000.50555,
          1000.5537 , 1000.5964 ],
         [ 996.22766,  996.19403,  996.16754, ..., 1000.52747,
          1000.54846, 1000.59247],
         [ 996.2059 ,  996.17035,  996.1449 , ..., 1000.53094,
          1000.54095, 1000.53094],
...
         [1003.3169 , 1003.2893 , 1003.2703 , ..., 1007.293  ,
          1007.8166 , 1007.92175],
         [1003.3373 , 1003.313  , 1003.3137 , ..., 1007.2611 ,
          1007.6963 , 1008.20807],
         [1003.3665 , 1003.3415 , 1003.3366 , ..., 1007.62726,
          1007.7819 , 1008.1645 ]],

        [[ 996.42914,  996.40393,  996.37726, ..., 1000.73865,
          1000.76373, 1000.80444],
         [ 996.418  ,  996.3734 ,  996.3398 , ..., 1000.7841 ,
          1000.78503, 1000.79395],
         [ 996.38885,  996.3414 ,  996.3321 , ..., 1000.7881 ,
          1000.78564, 1000.7676 ],
         ...,
         [1003.2633 , 1003.24927, 1003.2213 , ..., 1007.33185,
          1007.8745 , 1007.99817],
         [1003.3032 , 1003.2904 , 1003.2688 , ..., 1007.30493,
          1007.757  , 1008.22833],
         [1003.3352 , 1003.3148 , 1003.30646, ..., 1007.7138 ,
          1007.8635 , 1008.2027 ]]]], dtype=float32)
Coordinates:
    XLONG     (south_north, west_east) float32 116.5 116.6 116.6 ... 130.2 130.3
    XLAT      (south_north, west_east) float32 20.89 20.89 20.89 ... 32.78 32.78
    XTIME     (Time) float64 0.0 60.0 120.0
  * Time      (Time) datetime64[ns] 2019-08-08T18:00:00 ... 2019-08-08T20:00:00
    datetime  (Time) datetime64[ns] 2019-08-08T18:00:00 ... 2019-08-08T20:00:00
  * key_0     (key_0) <U5 'case1' 'case2'
Dimensions without coordinates: south_north, west_east
Attributes:
    FieldType:    104
    MemoryOrder:  XY
    description:  sea level pressure
    units:        hPa
    stagger:      
    coordinates:  XLONG XLAT XTIME
    projection:   LambertConformal(stand_lon=123.0, moad_cen_lat=27.0, truela...

恭喜你完成了 WRF 后处理训练营第一关的学习材料，学会了 WRF 数据的读取、变量的提取，更重要的是如何高效合并数据、合并变量，以 **更低的代码量、更高的运行速率来解析 WRF 数据**。  

接下来让我们完成一点闯关题，巩固一下。

## 闯关题  

### STEP1：根据要求完成题目：  


1. 查看 WRF 数据读取后的属性，本次模拟的水平分辨率是？        
        
	A. 1km    B. 2km   C. 3km   D. 4km        
        

2. `wrfout_d01_2019-08-08_18_00_00` 表示的时间是？        
	A. 北京时间 2019 年 8 月 8 日 18 时        
	B. 北京时间 2019 年 8 月 8 日 10 时        
	C. 北京时间 2019 年 8 月 9 日 02 时        
	D. 北京时间 2019 年 8 月 9 日 06 时        

  
3. 以下哪种读取方式 **无法** 实现多个 WRF 数据的合并？        
	A. xr.open_mfdataset ()        
	B. salem.open_mf_wrf_dataset ()        

    
4. 如果想禁用 / 关闭 xarray，应该将 getvar 函数中的 meta 参数设置为？        
	A. True        
	B. False        


In [ ]:
# 填入你的答案
answer_1 = ''  # 如 answer_1 = 'A'
answer_2 = ''
answer_3 = ''
answer_4 = ''

### STEP2：将结果保存为 csv 文件  

直接运行下方这个代码 cell，别改它。

In [ ]:
# 生成 csv 作业答案文件

def save_csv(a1, a2, a3, a4):
    import pandas as pd
    df = pd.DataFrame(
        {"id": ["q1", "q2", "q3", "q4"], "answer": [a1, a2, a3, a4]})
    df.to_csv("answer_wrf_1.csv", index=None)


# 运行这个cell,生成答案文件；该文件在左侧文件树project工作区下，你可以自行右击下载或者读取查看
save_csv(answer_1, answer_2, answer_3, answer_4)

### STEP3: 提交 csv 文件，获取分数结果  


你的 csv 答案文件已经准备完毕了，最后让我们提交答案文件，看看是否正确。  

提交方法：  

**1、拷贝提交 token**  

去对应关卡的 [提交页面](https://www.heywhale.com/home/competition/64478fec113e81a18dc70cd1/submit)，找到对应关卡，看到了你的 token 嘛？  

拷贝它。  

记得：每个关卡的 token 不一样。  

**2、下方 cell 里，拿你拷贝的 token 替换掉 XXXXXXX， 然后 Ctrl+Enter 运行 。**

In [ ]:
# 运行这个Cell 下载提交工具

!wget -nv -O heywhale_submit https://cdn.kesci.com/submit_tool/v4/heywhale_submit&&chmod +x heywhale_submit

# 运行提交工具
# 把下方XXXXXXX替换为你的 Token
# 改完看起来像是：!./heywhale_submit -token 586eeef71cb92941 -file answer_wrf_1.csv

!./heywhale_submit -token XXXXXXX -file answer_wrf_1.csv  # 替换XXXXXXX；注意不可增减任何空格或其他字符

运行成功、显示提交完成后，即可去 [提交页面](https://www.heywhale.com/home/competition/64478fec113e81a18dc70cd1/submit) 看成绩。满分即可进入下一关。  

没有成功也不怕，看下报错信息，对照《💡 [【提交出错】常见提交错误与排查建议》](https://www.heywhale.com/home/competition/forum/645a264a5ce8710c5bafb933) 帮助帖排查下，改完重新提交咯。

## ⚡ 下一关预告：  

下一关我们将介绍 WRF 模拟的降雨量的计算及空间绘制，实现初步的 WRF 模拟数据处理与可视化